engines/hypno/scene.cpp

```diff
@@ -0,0 +1,371 @@
+	"GS_SWITCH1",
+	"GS_SWITCH2",
+	"GS_SWITCH3",
+	"GS_SWITCH4",
+	"GS_SWITCH5",
+	"GS_SWITCH6",
+	"GS_SWITCH7",
+	"GS_SWITCH8",
+	"GS_SWITCH9",
+	"GS_SWITCH10",
+	"GS_SWITCH11",
+	"GS_SWITCH12",
+	"GS_COMBATJSON",
+	"GS_COMBATLEVEL",
+	"GS_PUZZLELEVEL",
+	NULL};
+
+void HypnoEngine::parseScene(const Common::String &prefix, const Common::String &filename) {
+	debugC(1, kHypnoDebugParser, "Parsing %s", filename.c_str());
+	Common::String name = convertPath(filename);
+	if (!prefix.empty())
+		name = prefix + "/" + name;
+	Common::File test;
+	assert(test.open(name.c_str()));
+
+	const uint32 fileSize = test.size();
+	char *buf = (char *)malloc(fileSize + 1);
+	test.read(buf, fileSize);
+	test.close();
+	buf[fileSize] = '\0';
+	parse_mis(buf);
+	Level level;
+	level.scene.prefix = prefix;
+	level.scene.hots = *g_parsedHots;
+	_levels[name] = level;
+	free(buf);
+}
+
+void HypnoEngine::resetSceneState() {
+	uint32 i = 0;
+	while (sceneVariables[i]) {
+		_sceneState[sceneVariables[i]] = 0;
+		i++;
+	}
+}
+
+bool HypnoEngine::checkSceneCompleted() {
+	return _sceneState["GS_LEVELCOMPLETE"];
+}
+
+// Hotspots
+
+void HypnoEngine::clickedHotspot(Common::Point mousePos) {
+	Hotspots *hots = stack.back();
+	Hotspot selected;
+	bool found = false;
+	int rs = 100000000;
+	int cs = 0;
+	for (Hotspots::const_iterator it = hots->begin(); it != hots->end(); ++it) {
+		const Hotspot h = *it;
+		if (h.type != MakeHotspot)
+			continue;
+
+		cs = h.rect.width() * h.rect.height();
+		if (h.rect.contains(mousePos)) {
+			if (cs < rs) {
+				selected = h;
+				found = true;
+				rs = cs;
+			}
+		}
+	}
+	if (found) {
+		if (selected.smenu) {
+			assert(!selected.smenu->empty());
+			_nextHotsToAdd = selected.smenu;
+		}
+
+		for (Actions::const_iterator itt = selected.actions.begin(); itt != selected.actions.end(); ++itt) {
+			Action *action = *itt;
+			if (typeid(*action) == typeid(ChangeLevel))
+				runChangeLevel((ChangeLevel *)action);
+			if (typeid(*action) == typeid(Escape))
+				runEscape((Escape *)action);
+			else if (typeid(*action) == typeid(Cutscene))
+				runCutscene((Cutscene *)action);
+			else if (typeid(*action) == typeid(Play))
+				runPlay((Play *)action);
+			else if (typeid(*action) == typeid(WalN))
+				runWalN((WalN *)action);
+			else if (typeid(*action) == typeid(Global))
+				runGlobal((Global *)action);
+			else if (typeid(*action) == typeid(Talk))
+				runTalk((Talk *)action);
+			else if (typeid(*action) == typeid(Quit))
+				runQuit((Quit *)action);
+			else if (typeid(*action) == typeid(Palette))
+				debugC(1, kHypnoDebugScene, "runPalette unimplemented");
+		}
+	}
+}
+
+bool HypnoEngine::hoverHotspot(Common::Point mousePos) {
+	Hotspots *hots = stack.back();
+	Hotspot selected;
+	bool found = false;
+	int rs = 100000000;
+	int cs = 0;
+	for (Hotspots::const_iterator it = hots->begin(); it != hots->end(); ++it) {
+		const Hotspot h = *it;
+		if (h.type != MakeHotspot)
+			continue;
+
+		cs = h.rect.width() * h.rect.height();
```

Comment:
`cs` declaration could be moved here for reducing its scope.

Reply:
Done.